tasks/Task.cpp

```diff
@@ -3,47 +3,355 @@
+
+    mWs->onMessage(
+        [&](variant<binary, string> data)
+        {
+            if (!holds_alternative<string>(data))
+                return;
+
+            if (!parseIncomingMessage(get<string>(data).c_str()))
+            {
+                return;
+            }
+
+            string type = decoder->getType();
+
+            // Ether the peerconnection is already created or we give an offer, creating one.
+            if (auto jt = mPeerConnectionMap.find("id"); jt != mPeerConnectionMap.end())
+            {
+                mPeerConnection = jt->second;
+            }
+            else if (type == "offer")
+            {
+                onOffer(mWs);
+            }
+            else
+            {
+                return;
+            }
+
+            if (type == "answer")
+            {
+                onAnswer();
+            }
+            else if (type == "candidate")
+            {
+                onCandidate();
+            }
+        });
+
+    // wss://signalserverhost?user=yourname
+    const string url = _websocket_server_name.get() + "?user=" + _local_peer_id.get();
+    mWs->open(url);
+    ws_future.get();
+}
+
+Task::Task(string const& name) : TaskBase(name) {}
+
+Task::~Task() {}
 
 /// The following lines are template definitions for the various state machine
 // hooks defined by Orocos::RTT. See Task.hpp for more detailed
 // documentation about them.
 
 bool Task::configureHook()
 {
-    if (! TaskBase::configureHook())
+    if (!TaskBase::configureHook())
         return false;
+
+    mConfig.iceServers.emplace_back(_ice_server.get());
+    mLocalPeerId = _local_peer_id.get();
+
+    configureWebSocket();
+
     return true;
 }
 bool Task::startHook()
 {
-    if (! TaskBase::startHook())
+    if (!TaskBase::startHook())
         return false;
     return true;
 }
 void Task::updateHook()
 {
+    RawPacket raw_packet;
+    if (_data_in.read(raw_packet) == RTT::NewData && !_remote_peer_id.get().empty())
+    {
+        shared_ptr<rtc::PeerConnection> peer_connection =
+            createPeerConnection(mWs, _remote_peer_id.get());
+
+        // We are the offerer, so create a data channel to initiate the process
+        string label = _data_channel_label.get();
+        LOG_INFO_S << "Creating DataChannel with label \"" << label << "\"" << std::endl;
+        shared_ptr<rtc::DataChannel> data_channel =
+            peer_connection->createDataChannel(label);
+
+        peer_connection->onDataChannel(
+            [&data_channel, raw_packet](shared_ptr<rtc::DataChannel> incoming)
+            {
+                data_channel = incoming;
+                vector<byte> data;
+                data.resize(raw_packet.data.size());
+                for(unsigned int i=0;i<raw_packet.data.size();i++)
+                {
+                    data[i] = static_cast<byte>(raw_packet.data[i]);
+                }
+                data_channel->send(&data.front(), data.size());
+            });
```

Comment:
You are supposed to create a data channel once, and use it forever. Not create one per message !

Reply:
> Start to look fine, but the signalling messages (in and out) do not match the format from rustysignal ...

yeah, I know. I'm still have doubt about that, so I decided to fix afterwards. I'm still getting confusing about how to parser the msg.